source/translated_table_model.ts

```diff
@@ -0,0 +1,197 @@
+      this.sourceToTranslatedIndices.push(index);
+    }
+    this.transactionArray = null;
+    this.transactionDepth = -1;
+  }
+
+  /**
+   * Marks the beginning of a transaction. In cases where a transaction is
+   * already being processed, then the sub-transaction gets consolidated into
+   * the parent transaction.
+   */
+  public beginTransaction(): void {
+    if(this.transactionArray === null) {
+      this.transactionArray = [];
+    }
+    this.transactionDepth += 1;
+  }
+
+  /** Ends a transaction. */
+  public endTransaction(): void {
+    if(this.transactionDepth === 0) {
+      this.dispatcher.dispatch(new Transaction(this.transactionArray));
+      this.transactionArray = null;
+    }
+    if(this.transactionDepth > -1) {
+      this.transactionDepth -= 1;
+    }
+  }
+
+  /**
+   * Moves a row.
+   * @param source - The index of the row to move.
+   * @param destination - The new index of the row.
+   * @throws RangeError - The index specified is not within range.
+   */
+  public moveRow(source: number, destination: number): void {
+    if(source >= this.rowCount || source < 0 || destination >= this.rowCount ||
+        destination < 0) {
+      throw new RangeError('The index specified is not within range.');
+    }
+    this.moveRowHelper(source, destination);
+  }
+
+  public get rowCount(): number {
+    return this.sourceTable.rowCount;
+  }
+
+  public get columnCount(): number {
+    return this.sourceTable.columnCount;
+  }
+
+  public get(row: number, column: number): any {
+    if(row >= this.rowCount || row < 0) {
+      throw new RangeError(
+        'Row or column are outside of the table\'s bounds.');
+    } 
+    return this.sourceTable.get(this.translatedToSourceIndices[row], column);
+  }
+
+  public connect(
+      slot: (operations: Operation) => void): Kola.Listener<Operation> {
+    return this.dispatcher.listen(slot);
+  }
+
+  private processOperation(operation: Operation) {
+    if(this.transactionArray === null) {
+      this.dispatcher.dispatch(operation);
+    } else {
+      this.transactionArray.push(operation);
+    }
+  }
+
+  private handleSourceOperation = (operation: Operation) => {
+    if(operation instanceof AddRowOperation) {
+      this.sourceAdd(operation);
+    } else if(operation instanceof MoveRowOperation) {
+      this.sourceMove(operation);
+    } else if(operation instanceof RemoveRowOperation) {
+      this.sourceRemove(operation);
+    } else if(operation instanceof UpdateOperation) {
+      this.sourceUpdate(operation);
+    } else if(operation instanceof Transaction) {
+      this.beginTransaction();
+      operation.operations.forEach(this.handleSourceOperation);
+      this.endTransaction();
+    }
+  }
+
+  private sourceAdd(operation: AddRowOperation) {
+    const rowCount = this.translatedToSourceIndices.length;
+    const originalIndices = {} as {[key: string]: number};
+    for(let index = operation.index; index < rowCount; ++index) {
```

Comment:
I don't think this is correct, do you have sufficient unit tests to verify that this works correctly? For example, consider a source table such as [1, 2, 3, 4] which is translated into [4, 3, 2, 1]. If the source adds 5 to the end ([1, 2, 3, 4, 5]), will the translation result in [4, 3, 2, 1, 5]?

    source = new ArrayTableModel();
    source.push([1]);
    source.push([2]);
    source.push([3]);
    source.push([4]);
    translation = new TranslatedTableModel(source);
    translation.move(3, 0);
    translation.move(3, 1);
    translation.move(3, 2);
    source.push([5]);

What is translation after this sequence of operations?

Reply:
I got the intended result [4, 3, 2, 1, 5] , but I'll separate the testSourceTableUpdates unit test into individual ones for each source operation.

Reply:
I got the intended result [4, 3, 2, 1, 5] but I'll add individual unit tests for each source operation. Is there anything about the implementation that seems incorrect or could be clarified?

Reply:
I found an error with the translation, where insertions at the source results in the first two elements of the translation being switched, so I'm currently debugging that issue.

Reply:
If this issue is resolved, click the "Resolve conversation" button.